ADR/0015-integration-service-two-phase-architecture.md

```diff
@@ -0,0 +1,108 @@
+multiple component builds can happen in quick succession, leading to a potential race
+condition between different Snapshots that are created for each of those builds.
+This would primarily manifest itself by two Snapshots getting promoted in quick
+succession but neither of them having the latest images contained within them.
+
+## Decision
+
+In order to protect against the race conditions, the integration service will leverage a
+two phase approach to testing. The phases will consist of a Component phase that will
+always be executed and the optional Composite phase which would come into play only when
+race conditions between component builds are detected.
+
+### Component phase
+
+When a single component image is built, the Integration Service tests the application
+by creating a Snapshot. All Components with their images from the Global Candidate List
+are included within the Snapshot and then the Component that was newly built is
+updated/overwritten to complete the Snapshot creation.
+
+After all test pipelines for the Snapshot finish successfully, the Integration service
+updates the Global Candidate List with the newly built Component image and checks if it
+can promote the Component Snapshot. If the Global Candidate List for other Components
+doesn't match the rest of the Component Snapshot contents, its status is marked as
+invalid and the testing goes into the Composite phase.
+Otherwise, the Component Snapshot is promoted according to user preferences.
+
+![](../diagrams/ADR-0015/component-phase.jpg)
+
+### Composite phase
+
+The Composite phase is used when the Global Candidate List changes while testing a
+Snapshot in the Component phase.
+
+The Composite phase exists to resolve a race condition when teams merge multiple PRs
+to multiple components of the same application at nearly the same time.
+When multiple components are built at the same time, the Integration Service tests
+the application by creating a composite Snapshot using multiple Components updated
+to use the newly built images.
+
+If all testing pipelines pass successfully, the Composite Snapshot is promoted
+according to user preferences.
+
+![](../diagrams/ADR-0015/composite-phase.jpg)
+
+## Consequences
+
+To illustrate the consequences of implementing the above approach,
+we can outline two scenarios, one with a happy path where only a single component
+is built at a time, and one with a race condition where two components are built
+in quick succession.
+
+### Happy path with a single component build
+
+In the happy path without race conditions, one PR merges to one component.
+1. The Component phase runs, and records results on the **Component** Snapshot.
+2. The Integration Service detects that the Snapshot is up-to-date by checking the
+Global Candidate List and skips the second phase of testing, marking the Snapshot as
+complete.
+3. The Global Candidate List is updated.
```

Comment:
According to above, the list is updated before marking as complete.

Reply:
Good catch, switched the order